File: lib/src/main/java/ua/naiksoftware/stomp/OkHttpConnectionProvider.java
package ua.naiksoftware.stomp;

import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.FlowableEmitter;
import okhttp3.Headers;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

/* package */ class OkHttpConnectionProvider implements ConnectionProvider {

    private static final String TAG = WebSocketsConnectionProvider.class.getSimpleName();

    private final String mUri;
    private final Map<String, String> mConnectHttpHeaders;
    private final OkHttpClient mOkHttpClient;

    private WebSocket openedSocked;

    private FlowableEmitter<? super String> mMessageEmitter;
    private final List<FlowableEmitter<? super LifecycleEvent>> mLifecycleEmitters = new ArrayList<>();

    OkHttpConnectionProvider(String uri, Map<String, String> connectHttpHeaders, OkHttpClient okHttpClient) {
        mUri = uri;
        mConnectHttpHeaders = connectHttpHeaders != null ? connectHttpHeaders : new HashMap<>();
        mOkHttpClient = okHttpClient;
    }


    @Override
    public Flowable<String> messages() {
        final Flowable<String> flowable = Flowable.<String>create(emitter -> mMessageEmitter = emitter, BackpressureStrategy.BUFFER);
        createWebSocketConnection();
        return flowable;
    }


    private void createWebSocketConnection() {

        if (openedSocked != null) {
            throw new IllegalStateException("Already have connection to web socket");
        }

        Request.Builder requestBuilder = new Request.Builder()
                .url(mUri);

        addConnectionHeadersToBuilder(requestBuilder, mConnectHttpHeaders);

        openedSocked = mOkHttpClient.newWebSocket(requestBuilder.build(),
                new WebSocketListener() {
                    @Override
                    public void onOpen(WebSocket webSocket, Response response) {
                        LifecycleEvent openEvent = new LifecycleEvent(LifecycleEvent.Type.OPENED);

                        TreeMap<String, String> headersAsMap = headersAsMap(response);

                        openEvent.setHandshakeResponseHeaders(headersAsMap);
                        emitLifecycleEvent(openEvent);
                    }

                    @Override
                    public void onMessage(WebSocket webSocket, String text) {
                        emitMessage(text);
                    }

                    @Override
                    public void onMessage(WebSocket webSocket, ByteString bytes) {
                        emitMessage(bytes.utf8());
                    }

                    @Override
                    public void onClosed(WebSocket webSocket, int code, String reason) {
                        emitLifecycleEvent(new LifecycleEvent(LifecycleEvent.Type.CLOSED));
                        openedSocked = null;
                    }

                    @Override
                    public void onFailure(WebSocket webSocket, Throwable t, Response response) {
                        emitLifecycleEvent(new LifecycleEvent(LifecycleEvent.Type.ERROR, new Exception(t)));
                    }

                    @Override
                    public void onClosing(final WebSocket webSocket, final int code, final String reason) {
                        webSocket.close(code, reason);
                    }
                }
        );
    }

    @Override
    public Flowable<Void> send(String stompMessage) {
        return Flowable.create(subscriber -> {
            if (openedSocked == null) {
                subscriber.onError(new IllegalStateException("Not connected yet"));
            } else {
                Log.d(TAG, "Send STOMP message: " + stompMessage);
                openedSocked.send(stompMessage);
                subscriber.onComplete();
            }
        }, BackpressureStrategy.BUFFER);
    }

    @Override
    public Flowable<LifecycleEvent> getLifecycleReceiver() {
        return Flowable.<LifecycleEvent>create(emitter->{
            synchronized (mLifecycleEmitters) {
                mLifecycleEmitters.add(emitter);
            }
        }, BackpressureStrategy.BUFFER);
    }

    private void closeSocket() {
        if (openedSocked != null) {
            openedSocked.close(1000, "");
            openedSocked = null;
        }
    }


    @Override
    public void disconnect() {
        closeSocket();
        mMessageEmitter = null;
        synchronized (mLifecycleEmitters) {
            mLifecycleEmitters.clear();
        }
    }

    private TreeMap<String, String> headersAsMap(Response response) {
        TreeMap<String, String> headersAsMap = new TreeMap<>();
        Headers headers = response.headers();
        for (String key : headers.names()) {
            headersAsMap.put(key, headers.get(key));
        }
        return headersAsMap;
    }

    private void addConnectionHeadersToBuilder(Request.Builder requestBuilder, Map<String, String> mConnectHttpHeaders) {
        for (Map.Entry<String, String> headerEntry : mConnectHttpHeaders.entrySet()) {
            requestBuilder.addHeader(headerEntry.getKey(), headerEntry.getValue());
        }
    }

    private void emitLifecycleEvent(LifecycleEvent lifecycleEvent) {
        synchronized (mLifecycleEmitters) {
            for (FlowableEmitter<? super LifecycleEvent> subscriber : mLifecycleEmitters) {
                subscriber.onNext(lifecycleEvent);
            }
        }
    }

    private void emitMessage(String stompMessage) {
        Log.d(TAG, "Emit STOMP message: " + stompMessage);
        FlowableEmitter<? super String> emitter = mMessageEmitter;
        if (emitter != null) {
            emitter.onNext(stompMessage);
        }
    }
}

File: lib/src/main/java/ua/naiksoftware/stomp/OkHttpConnectionProvider.kt
package ua.naiksoftware.stomp

import android.util.Log

import java.util.ArrayList
import java.util.HashMap
import java.util.TreeMap

import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableEmitter
import okhttp3.Headers
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString

internal class OkHttpConnectionProvider(private val mUri: String, connectHttpHeaders: Map<String, String>?,
                                        private val mOkHttpClient: OkHttpClient) : ConnectionProvider {

    companion object {
        private val TAG = WebSocketsConnectionProvider::class.java.simpleName
    }

    private val mConnectHttpHeaders: Map<String, String> = connectHttpHeaders ?: emptyMap()

    private var openedSocked: WebSocket? = null

    private var mMessageEmitter: FlowableEmitter<in String>? = null
    private val mLifecycleEmitters = ArrayList<FlowableEmitter<in LifecycleEvent>>()

    override val lifecycleReceiver: Flowable<LifecycleEvent>
        get() = Flowable.create({ emitter ->
            synchronized(mLifecycleEmitters) {
                mLifecycleEmitters.add(emitter)
            }
        }, BackpressureStrategy.BUFFER)


    override fun messages(): Flowable<String> {
        val flowable = Flowable.create<String>({ emitter -> mMessageEmitter = emitter }, BackpressureStrategy.BUFFER)
        createWebSocketConnection()
        return flowable
    }


    private fun createWebSocketConnection() {

        if (openedSocked != null) {
            throw IllegalStateException("Already have connection to web socket")
        }

        val requestBuilder = Request.Builder()
                .url(mUri)

        addConnectionHeadersToBuilder(requestBuilder, mConnectHttpHeaders)

        openedSocked = mOkHttpClient.newWebSocket(requestBuilder.build(),
                object : WebSocketListener() {
                    override fun onOpen(webSocket: WebSocket, response: Response) {
                        val openEvent = LifecycleEvent(LifecycleEvent.Type.OPENED)
                        emitLifecycleEvent(openEvent)
                    }

                    override fun onMessage(webSocket: WebSocket, text: String) {
                        emitMessage(text)
                    }

                    override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                        emitMessage(bytes.utf8())
                    }

                    override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                        emitLifecycleEvent(LifecycleEvent(LifecycleEvent.Type.CLOSED))
                        openedSocked = null
                    }

                    override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                        emitLifecycleEvent(LifecycleEvent(LifecycleEvent.Type.ERROR, Exception(t)))
                    }

                    override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                        webSocket.close(code, reason)
                    }
                }
        )
    }

    override fun send(stompMessage: String): Flowable<Void> {
        return Flowable.create({ subscriber ->
            if (openedSocked == null) {
                subscriber.onError(IllegalStateException("Not connected yet"))
            } else {
                Log.d(TAG, "Send STOMP message: " + stompMessage)
                openedSocked?.send(stompMessage)
                subscriber.onComplete()
            }
        }, BackpressureStrategy.BUFFER)
    }

    private fun closeSocket() {
        if (openedSocked != null) {
            openedSocked?.close(1000, "")
            openedSocked = null
        }
    }

    override fun disconnect() {
        closeSocket()
        mMessageEmitter = null
        synchronized(mLifecycleEmitters) {
            mLifecycleEmitters.clear()
        }
    }

    private fun addConnectionHeadersToBuilder(requestBuilder: Request.Builder, mConnectHttpHeaders: Map<String, String>) {
        for ((key, value) in mConnectHttpHeaders) {
            requestBuilder.addHeader(key, value)
        }
    }

    private fun emitLifecycleEvent(lifecycleEvent: LifecycleEvent) {
        synchronized(mLifecycleEmitters) {
            for (subscriber in mLifecycleEmitters) {
                subscriber.onNext(lifecycleEvent)
            }
        }
    }

    private fun emitMessage(stompMessage: String) {
        Log.d(TAG, "Emit STOMP message: " + stompMessage)
        val emitter = mMessageEmitter
        emitter?.onNext(stompMessage)
    }

}
